server/src/main/java/org/apache/iotdb/db/trigger/executor/TriggerFireVisitor.java

```diff
@@ -0,0 +1,292 @@
+    if (triggerNameToMeasurementList.isEmpty()) {
+      return TriggerFireResult.SUCCESS;
+    }
+    Map<String, Integer> measurementToSchemaIndexMap =
+        constructMeasurementToSchemaIndexMap(measurements, measurementSchemas);
+
+    Object[] values = node.getValues();
+    long time = node.getTime();
+    boolean hasFailedTrigger = false;
+    for (Map.Entry<String, List<String>> entry : triggerNameToMeasurementList.entrySet()) {
+      List<MeasurementSchema> schemas =
+          entry.getValue().stream()
+              .map(measurement -> measurementSchemas[measurementToSchemaIndexMap.get(measurement)])
+              .collect(Collectors.toList());
+      Tablet tablet = new Tablet(device, schemas);
+      tablet.addTimestamp(0, time);
+      for (String measurement : entry.getValue()) {
+        tablet.addValue(measurement, 0, values[measurementToSchemaIndexMap.get(measurement)]);
+      }
+      TriggerFireResult result = fire(entry.getKey(), tablet, context);
+      // Terminate if a trigger with pessimistic strategy messes up
+      if (result.equals(TriggerFireResult.TERMINATION)) {
+        return result;
+      }
+      if (result.equals(TriggerFireResult.FAILED_NO_TERMINATION)) {
+        hasFailedTrigger = true;
+      }
+    }
+    return hasFailedTrigger ? TriggerFireResult.FAILED_NO_TERMINATION : TriggerFireResult.SUCCESS;
+  }
+
+  @Override
+  public TriggerFireResult visitInsertTablet(InsertTabletNode node, TriggerEvent context) {
+    String device = node.getDevicePath().getFullPath();
+    String[] measurements = node.getMeasurements();
+    MeasurementSchema[] measurementSchemas = node.getMeasurementSchemas();
+    // group Triggers and measurements
+    Map<String, List<String>> triggerNameToMeasurementList =
+        constructTriggerNameToMeasurementListMap(node);
+    // return success if no trigger is found
+    if (triggerNameToMeasurementList.isEmpty()) {
+      return TriggerFireResult.SUCCESS;
+    }
+    Map<String, Integer> measurementToSchemaIndexMap =
+        constructMeasurementToSchemaIndexMap(measurements, measurementSchemas);
+
+    Object[] columns = node.getColumns();
+    BitMap[] bitMaps = node.getBitMaps();
+    long[] timestamps = node.getTimes();
+    boolean hasFailedTrigger = false;
+    for (Map.Entry<String, List<String>> entry : triggerNameToMeasurementList.entrySet()) {
+      List<MeasurementSchema> schemas =
+          entry.getValue().stream()
+              .map(measurement -> measurementSchemas[measurementToSchemaIndexMap.get(measurement)])
+              .collect(Collectors.toList());
+      Object[] columnsOfNewTablet =
+          entry.getValue().stream()
+              .map(measurement -> columns[measurementToSchemaIndexMap.get(measurement)])
+              .toArray();
+      BitMap[] bitMapsOfNewTablet =
+          entry.getValue().stream()
+              .map(measurement -> bitMaps[measurementToSchemaIndexMap.get(measurement)])
+              .toArray(BitMap[]::new);
```

Comment:
If the trigger's path pattern is like `root.**`, we may not bother to do this one by one, we just use all the columns together.